File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        mn = mx = ans = nums[0]
        
        for i in range(1, len(nums)):
            
            if nums[i] >= 0:
                mx, mn = max(mx * nums[i], nums[i]), min(mn * nums[i], nums[i])
            else:
                mx, mn = max(mn * nums[i], nums[i]), min(mx * nums[i], nums[i])
        
            ans = max(ans, mx)
        
        return ans

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int a = nums[0], b = nums[0], ans = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int n = nums[i];
            int prevMax = a;
            a = Math.max(a * n, Math.max(n, b * n));
            b = Math.min(prevMax * n, Math.min(b * n, n));
            System.out.println(a + " " + b);
            ans = Math.max(a, ans);
        }
        return ans;
    }
}